### ISSUE TITLE
`ObjectOmit` narrows `any` type to `{}`

### ISSUE THREAD

**Issue by OliverJAsh** [gcanti/typelevel-ts#19]
``` ts
import { ObjectOmit } from 'typelevel-ts';

type OmitUser<T> = ObjectOmit<T, 'user'>;

// Type is `{}`, expected `any`
type Result = OmitUser<any>;
```

Is there any way around this?

**Comment by tycho01** [comment]
hm. when removing string-based keys, doesn't it make sense that the output would be some kind of object?

**Comment by OliverJAsh** [comment]
I'm not sure 🤔 

For context, this behaviour is causing issues for me when I'm trying to use the [`withDefaults` HOC](https://github.com/gcanti/typelevel-ts#objectdiff), where the composed component doesn't have its prop types defined yet:

``` ts
class Foo extends React.Component<any> {}

const DefaultedFoo = withDefaults(Foo, { baz: 1 });

// Unexpected error: Property 'bar' does not exist on type 'IntrinsicAttributes & Pick<any, never> & Partial<{ baz: number; }> & { children?: ReactNode; }'.
<DefaultedFoo bar="foo" />;
```

This is because `ObjectDiff` uses `ObjectOmit`, and in this case it resolves to `Pick<any, never>`.

Although I am able to workaround the issue by widening the default props object to `any`:

``` ts
type Props = any;
class Foo extends React.Component<Props> {}

const defaults: Props = { baz: 1 }
const DefaultedFoo = withDefaults(Foo, defaults);

<DefaultedFoo bar="foo" />;
```

**Comment by tycho01** [comment]
Makes sense. `{}` admittedly is rather restrictive (no keys known to exist), whereas the intent in `any` was obviously to make it less restrictive, which seems to backfire now. Guess that does seem unfortunate.

**Comment by amir-arad** [comment]
I think design decisions in typescript made a turn for the worse :(
https://github.com/Microsoft/TypeScript/issues/20247

**Comment by IanYates** [comment]
Would Conditional Types - https://github.com/Microsoft/TypeScript/pull/21316 - help here?

**Comment by OliverJAsh** [comment]
@IanYates Maybe! Could we do something like `type OmitUser<user> = T extends any ? any : ObjectOmit<T, 'user'>`?

**Comment by tycho01** [comment]
I fear all types would extend `any` -- in that event I'd suggest flipping the check to `any extends T ?` -- I suppose that would actually hold only if `T` were to be `any` as well.
